pydantic_ai_slim/pydantic_ai/mcp.py

```diff
@@ -0,0 +1,160 @@
+
+    _client: ClientSession
+    _read_stream: MemoryObjectReceiveStream[JSONRPCMessage | Exception]
+    _write_stream: MemoryObjectSendStream[JSONRPCMessage]
+    _exit_stack: AsyncExitStack
+
+    @abstractmethod
+    @asynccontextmanager
+    async def client_streams(
+        self,
+    ) -> AsyncIterator[
+        tuple[MemoryObjectReceiveStream[JSONRPCMessage | Exception], MemoryObjectSendStream[JSONRPCMessage]]
+    ]:
+        """Create the streams for the MCP server."""
+        raise NotImplementedError('MCP Server subclasses must implement this method.')
+        yield
+
+    async def list_tools(self) -> list[ToolDefinition]:
+        """Retrieve tools that are currently active on the server.
+
+        Note:
+        - We don't cache tools as they might change.
+        - We also don't subscribe to the server to avoid complexity.
+        """
+        tools = await self._client.list_tools()
+        return [
+            ToolDefinition(
+                name=tool.name,
+                description=tool.description or '',
+                parameters_json_schema=tool.inputSchema,
+            )
+            for tool in tools.tools
+        ]
+
+    async def call_tool(self, tool_name: str, arguments: dict[str, Any]) -> CallToolResult:
+        """Call a tool on the server.
+
+        Args:
+            tool_name: The name of the tool to call.
+            arguments: The arguments to pass to the tool.
+
+        Returns:
+            The result of the tool call.
+        """
+        return await self._client.call_tool(tool_name, arguments)
+
+    async def __aenter__(self) -> MCPServer:
+        self._exit_stack = AsyncExitStack()
+
+        self._read_stream, self._write_stream = await self._exit_stack.enter_async_context(self.client_streams())
+        client = ClientSession(read_stream=self._read_stream, write_stream=self._write_stream)
+        self._client = await self._exit_stack.enter_async_context(client)
+
+        await self._client.initialize()
+        self.is_running = True
+        return self
+
+    async def __aexit__(
+        self, exc_type: type[BaseException] | None, exc_value: BaseException | None, traceback: TracebackType | None
+    ) -> bool | None:
+        await self._exit_stack.aclose()
+        self.is_running = False
+
+
+@dataclass
+class MCPSubprocessServer(MCPServer):
+    """An MCP server that runs a subprocess.
+
+    This class implements the stdio transport from the MCP specification.
+    See <https://modelcontextprotocol.io/docs/concepts/transports#standard-input%2Foutput-stdio> for more information.
+
+    Example:
+        ```python {py="3.10"}
+        from pydantic_ai import Agent
+        from pydantic_ai.mcp import MCPSubprocessServer
+
+        server = MCPSubprocessServer('python', ['-m', 'pydantic_ai_examples.mcp_server'])
+        agent = Agent('openai:gpt-4o', mcp_servers=[server])
+        ```
+    """
+
+    command: str
+    """The command to run."""
+
+    args: Sequence[str]
```

Comment:
Unlikely but this will allow `MCPSubprocessServer('python', '-m pydantic_ai_examples.mcp_server')` which will not work as expected.

Maybe enforce `list[str]`, or `list[str] | tuple[str, ...]`? Feel free to leave it as is.

Reply:
Should I use https://docs.python.org/3/library/shlex.html ?

Reply:
I'm not sure how `shlex` behaves on non Unix shells, and I think it's best to enforce a proper sequence of str instead of trying to parse `str` inputs

Reply:
So either we leave it as is if you think passing a bare string will _not_ be a common mistake, or change the type hint to `list[str] | tuple[str, ...]`?

Reply:
If someone get this mistake, I'll fix it.